### PR TITLE
Edit Project File

### PR DESCRIPTION
**Pull request by 333fred**
This is the initial progress for implementing #107. ~~In order to validate the changes, you'll need to be on a recent build of WPT or d15prerel, as the corresponding changes needed in msenv were checked in this week.~~ This adds a few new things:

* ~~I'm using COM Aggregation to implement `IVsTextBufferProvider` on the ProjectNode. Because the ProjectNode is stored in the RDT as the docdata for the project file, this means that when anything attempts to go look up the path to the project file, they'll get the ProjectNode, which can now return a TextBuffer when asked. Note that the `OpenProjectFile` capability needs to be present for this to work.~~
* I've added a new command to the right-click menu for the Project. The text will be "Edit _projectname_.csproj"
* ~~I've started work on wrapping the XML Editor and Editor Factory with our own implementations, so we can intercept calls made to `IOleCommandTarget.Exec`. This is what will allow us to get save and such working.~~

Things still to do:

* [x] Currently, I'm reading the project file from the disk. We need to figure out how to read it from CPS, or trigger a save before read. 
* [x] The editor wrapping needs to actually work. Right now, when using the Loaded Project File Factory, it displays a blank screen. This is likely because I'm not implementing some necessary interface, and the proper lifecycle calls aren't being made to the underlying xml editor.
* [ ] ~~Once the above two are working, we'll need to really think through the synchronization aspects here. It'll be very easy to get wrong.~~
    * We're not going to get this in time for RC. This will be worked on for RC.2.
* [x] VB Support (which should be relatively simple).

@dotnet/project-system @srivatsn @BillHiebert 

Update: While this solution was mostly working, getting the editor status showing correctly (ie the star in the window, and the changed line markers on the side of the buffer) wasn't achievable by RC. I've moved to a much simpler solution for now, which is copying the project xml to a temp file on right-click, and overwriting the xml on save. This has a major disadvantage compared to the original approach: simply opening the project file via File->Open isn't supported. We'll continue to work on the original approach for RC.2.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/LoadedProjectFileEditorFactory.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+using System.Runtime.InteropServices;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Editor
+{
+    [Guid(EditorFactoryGuid)]
+    class LoadedProjectFileEditorFactory : IVsEditorFactory
+    {
+        public const string EditorFactoryGuid = "da07c581-c7b4-482a-86fe-39aacfe5ca5c";
+        public static readonly Guid XmlEditorFactoryGuid = new Guid("{fa3cd31e-987b-443a-9b81-186104e8dac1}");
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/LoadedProjectFileEditorFactory.cs]: This needs to be refactored into a common area, other things shouldn't be depending on the LoadedProjectFileEditorFactory for this Guid. It's also defined elsewhere.

**Comment by 333fred** [comment]
The editor now supports saving of the project file. There's no concept of synchronization yet, so if something updates the project model, the editor isn't updated, and saving will overwrite the new changes on the disk.

**Comment by 333fred** [comment]
This is ready for review. Tagging @srivatsn, @BillHiebert, @dotnet/project-system.

/cc @MattGertz for approval. This change will only affect projects loaded with this project system, and won't run any code until the user hits `Edit <projname>.csproj/vproj`.

/cc @lifengl for the MSBuild access code.

**Comment by MattGertz** [comment]
Approved pending tests.

**Comment by srivatsn** [comment]
> using Microsoft.VisualStudio.Editor;

Copyright header

---
Refers to: src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/Input/Commands/EditCsprojCommand.cs:1 in 004b04b. [](commit_id = 004b04b176103b9f6575d2a5390d3770ec050da5, deletion_comment = False)

File: src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/Input/Commands/EditCsprojCommand.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.VisualStudio.ProjectSystem.Utilities;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using System;
+using System.ComponentModel.Composition;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    [ProjectCommand(ManagedProjectSystemPackage.DotnetProjectSystemCommandSet, ManagedProjectSystemPackage.EditProjectFileCmdId)]
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/Input/Commands/EditCsprojCommand.cs]: >DotnetProject [](start = 48, length = 13)

We should call this ManagedProjectSystemCommandSet to be consisten in the usage of managed vs dotnet

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Menus.vsct [hunk 3]
```diff
@@ -45,20 +41,44 @@
       </Button>
     </Buttons>
   </Commands>
-    
+
+  <Commands package="PackageGuidString">
+    <Groups>
+      <Group guid="guidDotnetProjectSystemCommandSet" id="cmdidEditProjFileGrp" priority="0x200">
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Menus.vsct]: >dDotnetP [](start = 22, length = 8)

Managed...

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            {
+                _fileSystem.WriteAllText(_unconfiguredProject.FullPath, savedText);
+                return Task.CompletedTask;
+            }));
+        }
+
+        private async Task<string> GetFile()
+        {
+            var projectXml = await _msbuildAccessor.GetProjectXml(_unconfiguredProject).ConfigureAwait(false);
+            var tempFileName = $"{_fileSystem.GetTempFileName()}.{FileExtension}";
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs]: Do we ever clean out this temp file?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/IMsBuildAccessor.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        [ImportingConstructor]
+        public MsBuildAccessor(IProjectLockService projectLockService)
+        {
+            _projectLockService = projectLockService;
+        }
+
+        public async Task<string> GetProjectXml(UnconfiguredProject unconfiguredProject)
+        {
+            var configuredProject = await unconfiguredProject.GetSuggestedConfiguredProjectAsync().ConfigureAwait(false);
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/IMsBuildAccessor.cs]: @lifengl? I just assumed that would give us the active based on this:https://github.com/Microsoft/VSProjectSystem/blob/master/doc/automation/finding_CPS_in_a_VS_project.md

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs (hunk 4 above)
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs]: Not currently.

**Comment by srivatsn** [comment]
> using Microsoft.VisualStudio.TextManager.Interop;

Copyright header

---
Refers to: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IVsPersistDocDataFactory.cs:1 in a239270. [](commit_id = a2392702b56986ffe63276e85c69b659dc02ad2d, deletion_comment = False)

**Comment by srivatsn** [comment]
> using System;

Copyright header

---
Refers to: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Utilities/TestShellUtilitiesHelper.cs:1 in a239270. [](commit_id = a2392702b56986ffe63276e85c69b659dc02ad2d, deletion_comment = False)

**Comment by srivatsn** [comment]
Looks good modulo the cleaning up of temp files.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs (hunk 4 above)
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs]: Implemented. The temp file is now cleaned on window close.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+            // Clean up the caption of the window
+            Verify.HResult(frame.SetProperty((int)__VSFPROPID5.VSFPROPID_OverrideCaption, caption));
+
+            // When the document is closed, clean up the file on disk
+            var fileCleanupListener = new EditProjectFileCleanupFrameNotifyListener(projPath, _fileSystem);
+            Verify.HResult(frame.SetProperty((int)__VSFPROPID.VSFPROPID_ViewHelper, fileCleanupListener));
+
+            // Ensure that the window is not reopened when the solution is closed
+            Verify.HResult(frame.SetProperty((int)__VSFPROPID5.VSFPROPID_DontAutoOpen, true));
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs]: If the vs is closed with the window open, would it still prevent the editor from opening when VS is opened?

Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs]: Yes

Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs]: If the solution is ever closed for any reason, this window won't reappear.

**Comment by nguerrera** [comment]
👏

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs (hunk 4 above)
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractEditProjectFileCommand.cs]: This also works if we close the solution or just VS as a whole.

**Pull request merged**